### PR TITLE
[SYCL] Add support for SYCL_EXTERNAL on device_global variables

### PR DESCRIPTION
**Pull request by Fznamznon**
Device global spec declares that it is possible to apply SYCL_EXTERNAL
macro to `device_global` variables. This patch modifies `sycl_device`
attribute which is used to implement SYCL_EXTERNAL macro so it is
possible to apply SYCL_EXTERNAL macro to `device_global` variables and
it is not allowed to apply it to any other variables.

### CODE REVIEW

File: clang/lib/Sema/Sema.cpp [hunk 1]
```diff
@@ -1727,6 +1727,11 @@ class DeferredDiagnosticsEmitter
               << Sema::KernelConstStaticVariable;
           return;
         }
+        if (!VD->hasInit() &&
```
Comment (premanandrao) [inline clang/lib/Sema/Sema.cpp:1730]: What is the hasInit check for?

Comment (Fznamznon) [reply clang/lib/Sema/Sema.cpp:1730]: It is needed to differentiate between a declaration of device_global variable and definition of it. The intention is to diagnose when externally defined device_global not marked with SYCL_EXTERNAL macro is referenced from SYCL device code. We have the similar diagnostic for SYCL_EXTERNAL and external functions.

**Comment by Fznamznon** [comment]
The following XPassed tests in `ESIMD Emu LLVM Test Suite (pull_request_target)` run:
```
********************
Unexpectedly Passed Tests (5):
  SYCL :: api/bin_and_cmp_ops_heavy.cpp
  SYCL :: api/simd_memory_access.cpp
  SYCL :: api/simd_view_select_2d_fp.cpp
  SYCL :: api/slm_gather_scatter_heavy.cpp
  SYCL :: api/unary_ops_heavy.cpp
```
seem to be a common problem across several PRs and therefore not related to this particular PR.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -11724,6 +11724,9 @@ def err_sycl_restrict : Error<
   "|use a const static or global variable that is neither zero-initialized "
   "nor constant-initialized"
   "}0">;
+def err_sycl_external_global : Error<
+  "SYCL device code cannot reference an external device_global variable not "
```
Comment (elizabethandrews) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:11728]: I'm not a huge fan of this diagnostic. WDYT about 

`invalid reference to device_global variable; external device_global variable must be marked with SYCL_EXTERNAL macro `

@premanandrao thoughts?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 3]
```diff
@@ -11755,9 +11758,12 @@ def err_sycl_function_attribute_mismatch : Error<
...
 def err_sycl_x_y_z_arguments_must_be_one : Error<
   "all %0 attribute arguments must be '1' when the %1 attribute argument is '0'">;
-def err_sycl_attribute_internal_function
+def err_sycl_attribute_internal_decl
     : Error<"%0 attribute cannot be applied to a "
-            "static function or function in an anonymous namespace">;
+            "static %select{function|variable}1 or %select{function|variable}1 "
+            "in an anonymous namespace">;
+def err_sycl_attribute_not_device_global
+    : Error<"%0 attribute can only appear on 'device_global' variables">;
```
Comment (elizabethandrews) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:11766]: `appear on` sounds odd. I think `can only be applied to` sounds better.

File: clang/lib/Sema/Sema.cpp [hunk 4]
```diff
@@ -1715,6 +1715,11 @@ class DeferredDiagnosticsEmitter
...
+            S.isTypeDecoratedWithDeclAttribute<SYCLGlobalVariableAllowedAttr>(
```
Comment (elizabethandrews) [inline clang/lib/Sema/Sema.cpp:1719]: Shouldn't this be `SYCLDeviceGlobalAttr` . `SYCLGlobalVariableAllowedAttr` can be used to mark other types in the future.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Sounds good to me. Thanks.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: LGTM, I've changed it.

File: clang/lib/Sema/Sema.cpp (hunk 4 above)
Comment (Fznamznon) [reply clang/lib/Sema/Sema.cpp]: I think you are right. I've made the change.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: I like it.  I would also suggest:
``invalid reference to external 'device_global' variable; such variables must be marked with SYCL_EXTERNAL macro``

File: clang/lib/Sema/Sema.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/lib/Sema/Sema.cpp]: This sounds right.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Is there a test for this diagnostic on a function?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: I'm not a huge fan of 'such variables'. I think  `invalid reference to external 'device_global' variable; variable must be marked with SYCL_EXTERNAL macro` suffices. I don't mind the existing diagnostic as well.

Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Fine by me.

File: clang/include/clang/Basic/Attr.td [hunk 5]
```diff
@@ -1247,18 +1247,18 @@ def CUDAShared : InheritableAttr {
...
 
+def GlobalStorageNonLocalVar : SubsetSubject<Var,
+                                             [{S->hasGlobalStorage() &&
+                                               !S->isLocalVarDeclOrParm()}],
+                                             "global variables">;
+
 def SYCLDevice : InheritableAttr {
   let Spellings = [GNU<"sycl_device">];
-  let Subjects = SubjectList<[Function]>;
+  let Subjects = SubjectList<[Function, GlobalStorageNonLocalVar]>;
```
Comment (elizabethandrews) [inline clang/include/clang/Basic/Attr.td:1257]: Sorry I didn't notice this earlier but does a diagnostic trigger while parsing the attribute when it doesn't satisfy the conditions for  `GlobalStorageNonLocalVar`

Comment (Fznamznon) [reply clang/include/clang/Basic/Attr.td:1257]: Yes, please see test [SemaSYCL/device_global_external.cpp](https://github.com/intel/llvm/pull/5963/files#diff-a79edb2ad29246c9fc79e3f70de77dfe2ad5539333ae87a5222334c489e5e843) on line 36. On attempt to apply the attribute to a local variable the diagnostic triggers.

**Review by elizabethandrews — APPROVED** [review]

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: @premanandrao , could you please elaborate a bit what exactly do you mean? `SYCL_EXTERNAL` macro can be legally applied to functions.

Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Sorry, looks like my comment got applied at the wrong place.
I meant for the previous diagnostic: ``err_sycl_external_global`` - if there were test cases where the '%1' argument is a function.   And to answer my own question, looks like they already exist and your changes don't affect it, just generlaizes them. 

So all set.  :-)

**Review by premanandrao — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by Fznamznon** [comment]
@intel/llvm-gatekeepers , I think this one is ready for merge.

**Pull request merged**